Link tracking

lina-brown opened the pull request:
Branch with several changes to allow us to track clicks on the links provided in the featured tool block by enabling an external service in combination with JS data attributes to tag the links that we wish to track.

lina-brown committed:
adding JS to test tracking links

lina-brown committed:
updating the format of the JS to be able to minify it correctly

lina-brown committed:
adding the new JS for tracking links to the renderer

lina-brown committed:
minified JS to be used by the plugin

lina-brown committed:
adding DB schema for link tracking table

lina-brown committed:
adding services class to be able to track the clicks

lina-brown committed:
adding external class for doing DB operations

lina-brown committed:
bump version to add DB and service components

lina-brown committed:
adding upgrade file to build the DB table for tracking

lina-brown committed:
adding information about link tracking to readme

lina-brown committed:
fixing spelling error

lina-brown committed:
fixing DB field config

lina-brown committed:
description message was unclear

lina-brown committed:
renaming a method to match convention

lina-brown committed:
change to regular JS needed change in minified JS

lina-brown committed:
updating the method name in one more spot

lina-brown committed:
fix for double logging

lina-brown committed:
adding service as built in rather than manually set up'

lina-brown committed:
removing length limit on link name

lina-brown committed:
last few fixes from codechecker

lina-brown committed:
few more spacing fixes

lina-brown committed:
removed a line added to wrong file

lina-brown committed:
had duplicate copyright blocks

lina-brown committed:
add validation for the data- attributes

lina-brown committed:
adding error messages for validation into lang file

lina-brown committed:
updating use case notes

lina-brown committed:
removing JS that I think is the culprit

lina-brown committed:
codechecker + docblock fixes

lina-brown committed:
missing period

lina-brown committed:
array syntax fix

lina-brown committed:
ended up w extra comma

haietza commented on README.md:
```diff
@@ -2,7 +2,7 @@
 Featured tool is a Moodle block plugin that allows admins to easily display information/announcements to teachers.
 
-The featured tool can be configured to cycle between multiple sets of information. Admins can enter a subtitle, thumbnail, and the content to display to the users when they click on a button. There are currently three tools to hold this information, and the featured tool will decide which to show randomly any time the dashboard is loaded. This information in the featured tool is configured only t ...
+The featured tool can be configured to cycle between multiple sets of information. Admins can enter a subtitle, thumbnail, and the content to display to the users when they click on a button. There are currently three tools to hold this information, and the featured tool will decide which to show randomly any time the dashboard is loaded. This information in the featured tool is configured only t ...
```
spelling error: meaninggul

haietza replied:
I think the instructions about the html attributes should be down in the configuration section, where there are other instructions about entering/formatting content.

lina-brown replied:
Got it - I'll move them! Thanks!

lina-brown replied:
This is done!

haietza commented on db/services.php:
```diff
@@ -0,0 +1,47 @@
+defined('MOODLE_INTERNAL') || die();
+
+$functions = [
+         'block_featured_tool_get_clicked_link' => [
```
Tabs should be 4 spaces, not 8; fix all impacted lines

haietza reviewed (changes requested):
Please also add information about the link tracking code to the text area help instructions: 
![Screenshot 2025-05-05 at 9 38 37 AM](https://github.com/user-attachments/assets/1a01bc7f-e15d-449f-a6aa-0246e009b908)

lina-brown committed:
I need to learn how to spell

lina-brown committed:
fix spacing in services class

lina-brown committed:
Updating help messages

lina-brown committed:
one more update to help text

lina-brown committed:
remove extra line break

lina-brown commented:
Thanks @haietza!

I have made the updated
1. Updated spelling in the readme
2. Fixed spacing in service.php and ran it back through codechecker
3. Added help text about how to make links trackable:

Full content to show to the user in the third featured tool. Displayed when user clicks on the third featured tool's button. Admins can enable link click tracking on featured tool links by adding the following HTML attributes to the anchor tag of the link: data-action="trackable" and data-name="name of link" through Tools > Source Code.'

Let me know if this needs to be adjusted further!

Thank you!

haietza reviewed (changes requested):
Just one more change on the README and I think we'll be good to go here. Thank you!

lina-brown committed:
moving link tracking info into the config section

lina-brown commented:
@haietza thank you! I have moved that blurb about how to track the links down to the config section of the readme!
